In [122]:
import tensorflow as tf

In [123]:
import pandas
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

In [124]:
# load cleaned dataset from ML notebook

X = pandas.read_csv('houseindependentfeature(x)')
Y = pandas.read_csv('housedependentfeature(y)')


# Splitting the data

In [125]:
from sklearn.model_selection import train_test_split,cross_val_score
x_train,x_test,y_train,y_test=train_test_split(X,Y,random_state=42)

In [126]:
x_train[:70]

0         1         2         3         4         5         6  \
1023  0.733315 -1.738121 -1.288571  0.328355 -0.289300 -1.409504  0.545992   
810   0.171142  1.058592 -1.074875 -1.140992  0.783542  0.787835 -1.466897   
1384 -1.680940 -0.258180  0.303110 -0.390770 -0.032945  0.334321 -0.591152   
626  -0.397531  1.321134  1.882836  0.052593 -0.996835 -1.764419 -0.434382   
813  -0.380443  0.151356  0.029064 -1.517830  0.495264 -1.394779 -0.201222   
...        ...       ...       ...       ...       ...       ...       ...   
462  -0.935842  1.060208  0.189800  0.196797 -0.762822  0.078529  0.053367   
129  -0.558280 -0.069805 -0.174619 -0.733331  0.331690 -0.182541  0.397045   
845   0.453591  2.545169 -0.336960  0.831364  0.603042 -0.269086  0.940367   
331  -0.461433 -0.110423 -0.815094 -0.816007 -1.097266  0.403030 -1.506469   
323  -0.418476 -0.422867 -0.614647 -0.878834 -1.286271  0.327244 -1.313352   

             7         8         9  ...        60        61        62  \
1023  0.350813  1.582540  0.112235  ...  0.331780 -0.623809 -0.148167   
810  -0.579147  0.090923  1.859657  ... -1.114019  0.167277  1.628146   
1384 -0.483299  0.106842 -1.889094  ...  0.634613 -1.464963  0.792218   
626  -0.344336  1.282965 -0.969296  ...  1.144884 -0.539174 -0.327358   
813  -0.193453  0.015431  0.845715  ...  0.272943 -1.326986 -0.269487   
...        ...       ...       ...  ...       ...       ...       ...   
462   0.334417 -2.076686  1.360641  ... -0.515580 -1.402849  1.002390   
129  -1.350728 -0.420182 -0.185568  ... -0.360187  2.251332 -0.498293   
845  -0.084283  0.800700 -0.464561  ...  0.460056  0.924993  1.233828   
331  -0.915684  0.454093  0.142877  ...  0.121514 -0.378047 -0.291084   
323  -0.846136  0.685449  1.369361  ...  0.298386  2.276640 -0.874803   

            63        64        65        66        67        68        69  
1023 -0.100425  0.450739 -0.106682 -0.481805 -0.167332 -0.324589 -0.590076  
810  -1.215375  2.334370  1.723775  2.892024  5.525734 -3.868929 -7.503845  
1384  1.812654 -1.477052  1.139525 -0.747134  0.826388  0.453397 -0.205676  
626   0.580151 -0.140406  0.075984 -0.223638 -0.366087  0.323312 -0.003670  
813  -0.477935  0.611329 -0.610763 -1.096396  0.070089 -0.358060  0.430118  
...        ...       ...       ...       ...       ...       ...       ...  
462   0.724149  2.000881  0.110515 -1.014898 -1.332115 -0.583252 -0.803204  
129  -0.405133  3.034763  3.131808  1.887797 -2.557489  1.987330 -0.574755  
845   0.507943 -1.420205 -1.149577  0.859540  0.375909  0.173379 -0.841055  
331   0.281982  0.884172 -0.861789  0.120837  0.399189 -0.311039  0.330971  
323  -0.794736  0.565831  0.086812  1.065327 -4.400662 -3.364705 -2.226235  

[70 rows x 70 columns]

In [127]:
x_train.shape,x_test.shape

((1095, 70), (365, 70))

In [128]:
y_train.shape,y_test.shape

((1095, 1), (365, 1))

# Sequential Model structure development

In [129]:
model=tf.keras.models.Sequential([
                                 #column=70   ,,give column numbers in rows place for input_shape=(number of columns,). 
    tf.keras.layers.Flatten(input_shape=(70,)),
                          #128,no.of neurons in the 2nd layer.
    tf.keras.layers.Dense(128,activation='relu'),
                    #Dropout()Regularization technique to avoid overfitting .
    tf.keras.layers.Dropout(0.2),
    
])
    


In [130]:
model.layers

In [131]:
model.input_shape

(None, 70)

In [132]:
model.output_shape

(None, 128)

In [133]:
                                                  # Weight * x +1 Bias
model.summary()#input 28 * 28 = 784 ,100480/128=785(784 + 1 Bias)

Model: "sequential_25"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_15 (Flatten)         (None, 70)                0         
_________________________________________________________________
dense_45 (Dense)             (None, 128)               9088      
_________________________________________________________________
dropout_15 (Dropout)         (None, 128)               0         
Total params: 9,088
Trainable params: 9,088
Non-trainable params: 0
_________________________________________________________________


# Predictions

In [134]:
predictions=model.predict(x_train[:1095])
predictions

array([[0.2201306 , 0.        , 0.        , ..., 0.        , 0.9221962 ,
        0.2670672 ],
       [0.32530898, 0.047813  , 1.1816788 , ..., 0.18694389, 0.        ,
        0.        ],
       [0.        , 0.80018014, 0.        , ..., 0.        , 0.62191594,
        0.20271286],
       ...,
       [0.46654418, 0.30700964, 0.        , ..., 0.46799222, 0.        ,
        0.        ],
       [0.        , 1.1705772 , 0.35593474, ..., 0.50273883, 1.2107043 ,
        0.2718138 ],
       [0.        , 0.        , 0.        , ..., 0.66404927, 0.        ,
        0.        ]], dtype=float32)

In [135]:
predictions.shape

(1095, 128)

# .compile()

In [136]:
model.compile(optimizer='adam',
              
              loss=['mean_squared_error'])


Now adam and mse are incorporated in the model

# Fitting the model

In [137]:
model.fit(x_train,y_train,epochs=5)

Epoch 1/5
35/35 [==============================] - 0s 2ms/step - loss: 136.9728
Epoch 2/5
35/35 [==============================] - 0s 2ms/step - loss: 135.7511
Epoch 3/5
35/35 [==============================] - 0s 2ms/step - loss: 134.5360
Epoch 4/5
35/35 [==============================] - 0s 2ms/step - loss: 133.4500
Epoch 5/5
35/35 [==============================] - 0s 2ms/step - loss: 132.2816


# Evaluation on test set

In [92]:
model.evaluate(x_test,y_test,verbose=2)

12/12 - 0s - loss: 131.9427


131.9427490234375

# Comparision with ML model

RMSE score=131.9427 < 135.378   as in case of ML model.
We can see the RMSE error decreasing in this case of deep learning model.



# Cross validation

In [103]:
def baseline_model():
    # create model
    model=tf.keras.models.Sequential([
                            #column=70   ,,give column numbers in rows place for input_shape=(number of columns,). 
    tf.keras.layers.Flatten(input_shape=(70,)),
                          #128,no.of neurons in the 2nd layer.
    tf.keras.layers.Dense(128,activation='relu'),
                    #Dropout()Regularization technique to avoid overfitting .
    tf.keras.layers.Dropout(0.2),
    
])
    
  # Compile model

    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

In [104]:
# evaluate model
estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0)
kfold = KFold(n_splits=10)
results = cross_val_score(estimator, X, Y, cv=kfold)
print("CVS= %.2f MSE= (%.2f) " % (results.mean(), results.std()))

CVS= -5.84 MSE= (0.07) 
